subsys/mgmt/hawkbit/hawkbit.c

```diff
@@ -598,6 +683,98 @@ static void hawkbit_dump_deployment(struct hawkbit_dep_res *d)
 	LOG_DBG("%s=%s", "md5sum-http", l->md5sum_http.href);
 }
 
+#ifdef CONFIG_HAWKBIT_SET_SETTINGS_RUNTIME
```

Comment:
This looks incredibly complex, I was expecting something simpler, will something like the following achieve what you want?

```suggestion
#ifdef CONFIG_HAWKBIT_SET_SETTINGS_RUNTIME
static void hawkbit_config_server_addr_nolock(const char *const addr)
{
	if (!strcmp(addr, hb_cfg.server_addr)) {
		LOG_DBG("%s already correctly configured", "hawkbit/server_addr");
		return;
	}

	strncpy(hb_cfg.server_addr, addr, sizeof(hb_cfg.server_addr));
	LOG_DBG("configured %s: %s", "hawkbit/server_addr", hb_cfg.server_addr);
}

static void hawkbit_config_server_port_nolock(const uint16_t port)
{
	if (port == atoi(hb_cfg.server_port)) {
		LOG_DBG("%s already correctly configured", "hawkbit/server_port");
		return;
	}

	snprintf(hb_cfg.server_port, sizeof(hb_cfg.server_port), "%u", port);
	LOG_DBG("configured %s: %s", "hawkbit/server_port", hb_cfg.server_port);
}

#ifndef CONFIG_HAWKBIT_DDI_NO_SECURITY
static void hawkbit_config_server_auth_nolock(const char *const auth)
{
	if (!strcmp(auth, hb_cfg.ddi_security_token)) {
		LOG_DBG("%s already correctly configured", "hawkbit/ddi_token");
	}

	strncpy(hb_cfg.ddi_security_token, auth, sizeof(hb_cfg.ddi_security_token));
	LOG_DBG("configured %s: %s", "hawkbit/ddi_token", hb_cfg.ddi_security_token);
}
#endif /* CONFIG_HAWKBIT_DDI_NO_SECURITY */

int hawkbit_config_server(struct hawkbit_set_config *set_config)
{
	if (k_sem_take(&probe_sem, HAWKBIT_SET_SERVER_TIMEOUT) != 0) {
		LOG_DBG("Failed to configure server");
		return -EAGAIN;
	}

	if (set_config->addr != NULL) {
		hawkbit_config_server_addr_nolock(set_config->addr);
	}

	if (set_config->port != 0U) {
		hawkbit_config_server_port_nolock(set_config->port);
	}

#ifndef CONFIG_HAWKBIT_DDI_NO_SECURITY
	if (set_config->auth != NULL) {
		hawkbit_config_server_auth_nolock(set_config->auth);
	}
#endif /* CONFIG_HAWKBIT_DDI_NO_SECURITY */

	settings_save();

	k_sem_give(&probe_sem);

	return 0;
}
#endif /* CONFIG_HAWKBIT_SET_SETTINGS_RUNTIME */
```

Public functions like `hawkbit_set_server_addr` can call `hawkbit_config_server_addr_nolock` with the lock held, arg check can go into the internal functions if desired, they can be made to return error codes if necessary.

Reply:
Well I wanted to do the check if a value is already correctly configured before taking the lock, so that if nothing changes taking the lock is not required.

Reply:
I'd expect that the function is called only when there's something that needs to be changed, if an application is calling the config function repeatedly without the need to change anything then the application has to be fixed.

Reply:
I will remove the part where it will compare the old and new value

Reply:
@ycsin it's changed now